File: engine/src/registration/pointIndex.ts
/**
 * Uniform-grid spatial index and the two preprocessing steps ICP needs:
 * voxel downsampling and PCA normal estimation.
 *
 * A uniform grid, not a KD-tree, for the reason the cloud-to-cloud distance
 * plan in docs/BACKLOG.md gives: after voxel downsampling the points are close
 * to uniformly dense, so bucketing is O(n) to build, allocates three typed
 * arrays, and a query touches a fixed number of cells. A KD-tree would win only
 * on the pathological density variation the downsample already removed.
 *
 * Buckets are stored CSR-style (`cellStart` + `items`) rather than as an array
 * of arrays: a station scan downsamples to a few hundred thousand points, and
 * one small object per occupied cell is the difference between a smooth refine
 * and a garbage-collection stutter mid-iteration.
 */

import { symmetricEigen } from './linalg';

/** Cap on grid cells, so a tiny cell size on a large scene cannot blow up memory. */
const MAX_CELLS = 1 << 22;

export class PointGrid {
  readonly points: Float32Array;
  readonly cellSize: number;
  private readonly minX: number;
  private readonly minY: number;
  private readonly minZ: number;
  private readonly nx: number;
  private readonly ny: number;
  private readonly nz: number;
  private readonly cellStart: Int32Array;
  private readonly items: Int32Array;

  /**
   * @param points Flat xyz triples. Retained by reference, not copied.
   * @param cellSize Requested cell size; enlarged if the bounding box would
   *   need more than `MAX_CELLS` cells.
   */
  constructor(points: Float32Array, cellSize: number) {
    this.points = points;
    const count = Math.floor(points.length / 3);

    let minX = Infinity;
    let minY = Infinity;
    let minZ = Infinity;
    let maxX = -Infinity;
    let maxY = -Infinity;
    let maxZ = -Infinity;
    for (let i = 0; i < count; i++) {
      const x = points[i * 3];
      const y = points[i * 3 + 1];
      const z = points[i * 3 + 2];
      if (x < minX) {minX = x;}
      if (y < minY) {minY = y;}
      if (z < minZ) {minZ = z;}
      if (x > maxX) {maxX = x;}
      if (y > maxY) {maxY = y;}
      if (z > maxZ) {maxZ = z;}
    }
    if (count === 0) {
      minX = minY = minZ = 0;
      maxX = maxY = maxZ = 0;
    }

    const spanX = Math.max(maxX - minX, 1e-6);
    const spanY = Math.max(maxY - minY, 1e-6);
    const spanZ = Math.max(maxZ - minZ, 1e-6);
    let size = Math.max(cellSize, 1e-6);
    // Grow the cell until the lattice fits the budget. Each doubling cuts the
    // cell count by 8, so this terminates in a handful of steps.
    while (
      Math.ceil(spanX / size) * Math.ceil(spanY / size) * Math.ceil(spanZ / size) >
      MAX_CELLS
    ) {
      size *= 2;
    }

    this.cellSize = size;
    this.minX = minX;
    this.minY = minY;
    this.minZ = minZ;
    this.nx = Math.max(1, Math.ceil(spanX / size) + 1);
    this.ny = Math.max(1, Math.ceil(spanY / size) + 1);
    this.nz = Math.max(1, Math.ceil(spanZ / size) + 1);

    // Counting sort into CSR buckets: one pass to count, prefix sum, one pass
    // to place.
    const cellCount = this.nx * this.ny * this.nz;
    const counts = new Int32Array(cellCount + 1);
    const cellOf = new Int32Array(count);
    for (let i = 0; i < count; i++) {
      const cell = this.cellIndex(points[i * 3], points[i * 3 + 1], points[i * 3 + 2]);
      cellOf[i] = cell;
      counts[cell + 1]++;
    }
    for (let c = 0; c < cellCount; c++) {
      counts[c + 1] += counts[c];
    }
    this.cellStart = counts;
    this.items = new Int32Array(count);
    const cursor = counts.slice(0, cellCount);
    for (let i = 0; i < count; i++) {
      this.items[cursor[cellOf[i]]++] = i;
    }
  }

  private cellIndex(x: number, y: number, z: number): number {
    const ix = Math.min(this.nx - 1, Math.max(0, Math.floor((x - this.minX) / this.cellSize)));
    const iy = Math.min(this.ny - 1, Math.max(0, Math.floor((y - this.minY) / this.cellSize)));
    const iz = Math.min(this.nz - 1, Math.max(0, Math.floor((z - this.minZ) / this.cellSize)));
    return (iz * this.ny + iy) * this.nx + ix;
  }

  /**
   * Index of the closest indexed point within `maxDistance`, or -1.
   *
   * Bounded rather than unbounded on purpose: an ICP correspondence beyond the
   * gate is rejected anyway, and the bound is what keeps the number of visited
   * cells constant.
   */
  nearest(x: number, y: number, z: number, maxDistance: number): number {
    const ring = Math.max(1, Math.ceil(maxDistance / this.cellSize));
    const cx = Math.floor((x - this.minX) / this.cellSize);
    const cy = Math.floor((y - this.minY) / this.cellSize);
    const cz = Math.floor((z - this.minZ) / this.cellSize);

    let best = -1;
    let bestSq = maxDistance * maxDistance;
    for (let iz = Math.max(0, cz - ring); iz <= Math.min(this.nz - 1, cz + ring); iz++) {
      for (let iy = Math.max(0, cy - ring); iy <= Math.min(this.ny - 1, cy + ring); iy++) {
        const rowBase = (iz * this.ny + iy) * this.nx;
        const fromX = Math.max(0, cx - ring);
        const toX = Math.min(this.nx - 1, cx + ring);
        for (let ix = fromX; ix <= toX; ix++) {
          const cell = rowBase + ix;
          const end = this.cellStart[cell + 1];
          for (let slot = this.cellStart[cell]; slot < end; slot++) {
            const index = this.items[slot];
            const dx = this.points[index * 3] - x;
            const dy = this.points[index * 3 + 1] - y;
            const dz = this.points[index * 3 + 2] - z;
            const squared = dx * dx + dy * dy + dz * dz;
            if (squared < bestSq) {
              bestSq = squared;
              best = index;
            }
          }
        }
      }
    }
    return best;
  }

  /** Indices within `radius` of the query, appended to `out`, up to `limit`. */
  neighbors(x: number, y: number, z: number, radius: number, limit: number, out: number[]): void {
    out.length = 0;
    const ring = Math.max(1, Math.ceil(radius / this.cellSize));
    const cx = Math.floor((x - this.minX) / this.cellSize);
    const cy = Math.floor((y - this.minY) / this.cellSize);
    const cz = Math.floor((z - this.minZ) / this.cellSize);
    const radiusSq = radius * radius;

    for (let iz = Math.max(0, cz - ring); iz <= Math.min(this.nz - 1, cz + ring); iz++) {
      for (let iy = Math.max(0, cy - ring); iy <= Math.min(this.ny - 1, cy + ring); iy++) {
        const rowBase = (iz * this.ny + iy) * this.nx;
        for (let ix = Math.max(0, cx - ring); ix <= Math.min(this.nx - 1, cx + ring); ix++) {
          const cell = rowBase + ix;
          const end = this.cellStart[cell + 1];
          for (let slot = this.cellStart[cell]; slot < end; slot++) {
            const index = this.items[slot];
            const dx = this.points[index * 3] - x;
            const dy = this.points[index * 3 + 1] - y;
            const dz = this.points[index * 3 + 2] - z;
            if (dx * dx + dy * dy + dz * dz <= radiusSq) {
              out.push(index);
              if (out.length >= limit) {
                return;
              }
            }
          }
        }
      }
    }
  }
}

/**
 * Averages the points falling in each `voxelSize` cell down to one point.
 *
 * Averaging rather than picking a representative: a station scan is far denser
 * near the scanner than at range, and keeping an arbitrary member of each cell
 * leaves that density gradient in the residual, which biases the ICP fit toward
 * whatever is closest to the tripod.
 */
export function voxelDownsample(points: Float32Array, voxelSize: number): Float32Array {
  const count = Math.floor(points.length / 3);
  if (count === 0 || !(voxelSize > 0)) {
    return new Float32Array(0);
  }

  let minX = Infinity;
  let minY = Infinity;
  let minZ = Infinity;
  for (let i = 0; i < count; i++) {
    minX = Math.min(minX, points[i * 3]);
    minY = Math.min(minY, points[i * 3 + 1]);
    minZ = Math.min(minZ, points[i * 3 + 2]);
  }

  const sums = new Map<string, { x: number; y: number; z: number; n: number }>();
  for (let i = 0; i < count; i++) {
    const x = points[i * 3];
    const y = points[i * 3 + 1];
    const z = points[i * 3 + 2];
    if (!Number.isFinite(x) || !Number.isFinite(y) || !Number.isFinite(z)) {
      continue;
    }
    const key = `${Math.floor((x - minX) / voxelSize)},${Math.floor((y - minY) / voxelSize)},${Math.floor((z - minZ) / voxelSize)}`;
    const bucket = sums.get(key);
    if (bucket) {
      bucket.x += x;
      bucket.y += y;
      bucket.z += z;
      bucket.n++;
    } else {
      sums.set(key, { x, y, z, n: 1 });
    }
  }

  const out = new Float32Array(sums.size * 3);
  let write = 0;
  for (const bucket of sums.values()) {
    out[write++] = bucket.x / bucket.n;
    out[write++] = bucket.y / bucket.n;
    out[write++] = bucket.z / bucket.n;
  }
  return out;
}

/**
 * Per-point surface normals from the smallest principal component of the
 * neighborhood within `radius`.
 *
 * Orientation is left arbitrary. Point-to-plane ICP squares the plane residual,
 * so a flipped normal costs nothing, and consistently orienting them across a
 * scan is a much harder problem than the solver actually needs solved.
 *
 * Points with too few neighbors get a zero normal; `icp` treats that as "no
 * usable plane here" and skips the correspondence.
 */
export function estimateNormals(grid: PointGrid, radius: number, minNeighbors = 6): Float32Array {
  const points = grid.points;
  const count = Math.floor(points.length / 3);
  const normals = new Float32Array(count * 3);
  const neighborhood: number[] = [];

  for (let i = 0; i < count; i++) {
    const x = points[i * 3];
    const y = points[i * 3 + 1];
    const z = points[i * 3 + 2];
    grid.neighbors(x, y, z, radius, 48, neighborhood);
    if (neighborhood.length < minNeighbors) {
      continue;
    }

    let cx = 0;
    let cy = 0;
    let cz = 0;
    for (const index of neighborhood) {
      cx += points[index * 3];
      cy += points[index * 3 + 1];
      cz += points[index * 3 + 2];
    }
    const n = neighborhood.length;
    cx /= n;
    cy /= n;
    cz /= n;

    let xx = 0;
    let xy = 0;
    let xz = 0;
    let yy = 0;
    let yz = 0;
    let zz = 0;
    for (const index of neighborhood) {
      const dx = points[index * 3] - cx;
      const dy = points[index * 3 + 1] - cy;
      const dz = points[index * 3 + 2] - cz;
      xx += dx * dx;
      xy += dx * dy;
      xz += dx * dz;
      yy += dy * dy;
      yz += dy * dz;
      zz += dz * dz;
    }

    const eigen = symmetricEigen([xx, xy, xz, xy, yy, yz, xz, yz, zz], 3);
    const normal = eigen.vectors[2];
    const length = Math.hypot(normal[0], normal[1], normal[2]);
    if (length > 1e-12) {
      normals[i * 3] = normal[0] / length;
      normals[i * 3 + 1] = normal[1] / length;
      normals[i * 3 + 2] = normal[2] / length;
    }
  }

  return normals;
}

/**
 * Robust working diameter of a cloud: twice the 90th-percentile distance from
 * its median centre.
 *
 * The bounding box is the wrong scale reference for a station scan. Half the
 * points of a real scan sit within a few metres of the tripod while a thin tail
 * of long-range returns stretches the box to a hundred metres or more, so any
 * parameter derived from the box (a voxel size, a raster cell) ends up an order
 * of magnitude too coarse for the geometry that two stations actually share.
 */
export function robustExtent(points: Float32Array, maxSamples = 100_000): number {
  const count = Math.floor(points.length / 3);
  if (count === 0) {
    return 0;
  }
  const stride = Math.max(1, Math.ceil(count / maxSamples));
  const axes: number[][] = [[], [], []];
  for (let i = 0; i < count; i += stride) {
    for (let axis = 0; axis < 3; axis++) {
      axes[axis].push(points[i * 3 + axis]);
    }
  }
  const centre = axes.map(values => {
    const sorted = values.slice().sort((x, y) => x - y);
    return sorted[Math.floor(sorted.length / 2)];
  });

  const radii: number[] = [];
  for (let i = 0; i < axes[0].length; i++) {
    radii.push(Math.hypot(axes[0][i] - centre[0], axes[1][i] - centre[1], axes[2][i] - centre[2]));
  }
  radii.sort((x, y) => x - y);
  return Math.max(radii[Math.floor(radii.length * 0.9)] * 2, 1e-6);
}

File: wasm/pointcloud-parser/src/registration/point_index.rs
//! Uniform-grid spatial index plus the two preprocessing steps ICP needs:
//! voxel downsampling and PCA normal estimation.
//!
//! A uniform grid rather than a KD-tree: after voxel downsampling the points
//! are close to uniformly dense, so bucketing is O(n) to build and a query
//! touches a fixed number of cells. Buckets are CSR-style (`cell_start` +
//! `items`) rather than a Vec per cell — a station scan downsamples to a few
//! hundred thousand points and one allocation per occupied cell is pure waste.

use std::collections::HashMap;

use super::linalg::symmetric_eigen;

/// Hash of a cell coordinate. FxHash-style multiply-xor rather than the
/// default SipHash: the keys are three small integers, lookups happen tens of
/// millions of times per solve, and cryptographic strength buys nothing here.
#[derive(Default, Clone, Copy)]
pub struct CellHasher(u64);

impl std::hash::Hasher for CellHasher {
    fn finish(&self) -> u64 {
        self.0
    }
    fn write(&mut self, bytes: &[u8]) {
        for &byte in bytes {
            self.write_u64(byte as u64);
        }
    }
    fn write_i32(&mut self, value: i32) {
        self.write_u64(value as u32 as u64);
    }
    fn write_u64(&mut self, value: u64) {
        self.0 = (self.0 ^ value).wrapping_mul(0x517c_c1b7_2722_0a95);
    }
}

#[derive(Default, Clone, Copy)]
pub struct CellHasherBuilder;

impl std::hash::BuildHasher for CellHasherBuilder {
    type Hasher = CellHasher;
    fn build_hasher(&self) -> CellHasher {
        CellHasher(0xcbf2_9ce4_8422_2325)
    }
}

type CellKey = (i32, i32, i32);

/// Sparse uniform grid over occupied cells only.
///
/// A dense lattice was the obvious first implementation and it is the wrong one
/// for this data: a station scan's bounding box is set by a handful of
/// long-range returns (150 m on a real archive) while its useful structure sits
/// inside a few metres, so a lattice covering the box at the correspondence
/// gate's resolution needs billions of cells. Capping the cell count then
/// inflates the cell size until each one holds thousands of points and every
/// nearest-neighbour query degenerates into a linear scan — measured at 0.65 s
/// per ICP iteration before this changed. Hashing the occupied cells makes
/// empty space free, so the cell size can stay at the gate where it belongs.
pub struct PointGrid {
    pub points: Vec<f32>,
    cell_size: f64,
    /// Occupied cell -> its span in `items`.
    cells: HashMap<CellKey, (u32, u32), CellHasherBuilder>,
    items: Vec<u32>,
}

impl PointGrid {
    /// `points` is flat xyz triples, taken by value: an ICP run holds its grids
    /// for the whole solve, and a borrowed slice would make the cached scale
    /// levels in `icp.rs` self-referential for no gain.
    pub fn new(points: Vec<f32>, cell_size: f64) -> Self {
        let count = points.len() / 3;
        let size = cell_size.max(1e-6);

        // Sort point indices by cell, then hand each cell its contiguous span.
        let mut keyed: Vec<(CellKey, u32)> = Vec::with_capacity(count);
        for i in 0..count {
            keyed.push((
                cell_key(
                    points[i * 3] as f64,
                    points[i * 3 + 1] as f64,
                    points[i * 3 + 2] as f64,
                    size,
                ),
                i as u32,
            ));
        }
        keyed.sort_unstable_by_key(|entry| entry.0);

        let mut cells: HashMap<CellKey, (u32, u32), CellHasherBuilder> =
            HashMap::with_capacity_and_hasher(keyed.len() / 4 + 1, CellHasherBuilder);
        let mut items = Vec::with_capacity(count);
        let mut index = 0usize;
        while index < keyed.len() {
            let key = keyed[index].0;
            let start = items.len() as u32;
            while index < keyed.len() && keyed[index].0 == key {
                items.push(keyed[index].1);
                index += 1;
            }
            cells.insert(key, (start, items.len() as u32));
        }

        PointGrid {
            points,
            cell_size: size,
            cells,
            items,
        }
    }

    /// Visits every indexed point within `radius` of the query.
    fn for_each_near(
        &self,
        x: f64,
        y: f64,
        z: f64,
        radius: f64,
        mut visit: impl FnMut(usize, f64),
    ) {
        let ring = (radius / self.cell_size).ceil().max(1.0) as i32;
        let centre = cell_key(x, y, z, self.cell_size);
        let radius_sq = radius * radius;

        for iz in (centre.2 - ring)..=(centre.2 + ring) {
            for iy in (centre.1 - ring)..=(centre.1 + ring) {
                for ix in (centre.0 - ring)..=(centre.0 + ring) {
                    let Some(&(start, end)) = self.cells.get(&(ix, iy, iz)) else {
                        continue;
                    };
                    for &item in &self.items[start as usize..end as usize] {
                        let index = item as usize;
                        let dx = self.points[index * 3] as f64 - x;
                        let dy = self.points[index * 3 + 1] as f64 - y;
                        let dz = self.points[index * 3 + 2] as f64 - z;
                        let squared = dx * dx + dy * dy + dz * dz;
                        if squared <= radius_sq {
                            visit(index, squared);
                        }
                    }
                }
            }
        }
    }

    /// Index of the closest indexed point within `max_distance`, or `None`.
    ///
    /// Bounded on purpose: an ICP correspondence beyond the gate is rejected
    /// anyway, and the bound is what keeps the visited cell count constant.
    pub fn nearest(&self, x: f64, y: f64, z: f64, max_distance: f64) -> Option<usize> {
        let mut best = None;
        let mut best_sq = max_distance * max_distance;
        self.for_each_near(x, y, z, max_distance, |index, squared| {
            if squared < best_sq {
                best_sq = squared;
                best = Some(index);
            }
        });
        best
    }

    pub fn neighbors(
        &self,
        x: f64,
        y: f64,
        z: f64,
        radius: f64,
        limit: usize,
        out: &mut Vec<usize>,
    ) {
        out.clear();
        self.for_each_near(x, y, z, radius, |index, _| {
            if out.len() < limit {
                out.push(index);
            }
        });
    }
}

fn cell_key(x: f64, y: f64, z: f64, cell_size: f64) -> CellKey {
    (
        (x / cell_size).floor() as i32,
        (y / cell_size).floor() as i32,
        (z / cell_size).floor() as i32,
    )
}

/// Averages the points falling in each `voxel_size` cell down to one point.
///
/// Averaging rather than picking a representative: a station scan is far denser
/// near the scanner than at range, and keeping an arbitrary member of each cell
/// leaves that density gradient in the residual, biasing the fit toward
/// whatever is closest to the tripod.
pub fn voxel_downsample(points: &[f32], voxel_size: f64) -> Vec<f32> {
    let count = points.len() / 3;
    if count == 0 || !(voxel_size > 0.0) {
        return Vec::new();
    }
    let mut min = [f64::INFINITY; 3];
    for i in 0..count {
        for axis in 0..3 {
            min[axis] = min[axis].min(points[i * 3 + axis] as f64);
        }
    }

    let mut buckets: HashMap<(i64, i64, i64), (f64, f64, f64, u32)> = HashMap::new();
    for i in 0..count {
        let (x, y, z) = (
            points[i * 3] as f64,
            points[i * 3 + 1] as f64,
            points[i * 3 + 2] as f64,
        );
        if !x.is_finite() || !y.is_finite() || !z.is_finite() {
            continue;
        }
        let key = (
            ((x - min[0]) / voxel_size).floor() as i64,
            ((y - min[1]) / voxel_size).floor() as i64,
            ((z - min[2]) / voxel_size).floor() as i64,
        );
        let entry = buckets.entry(key).or_insert((0.0, 0.0, 0.0, 0));
        entry.0 += x;
        entry.1 += y;
        entry.2 += z;
        entry.3 += 1;
    }

    let mut out = Vec::with_capacity(buckets.len() * 3);
    // Sorted so the output order does not depend on hash iteration order:
    // a registration run has to be reproducible.
    let mut keys: Vec<_> = buckets.keys().copied().collect();
    keys.sort_unstable();
    for key in keys {
        let (sx, sy, sz, n) = buckets[&key];
        let n = n as f64;
        out.push((sx / n) as f32);
        out.push((sy / n) as f32);
        out.push((sz / n) as f32);
    }
    out
}

/// Evenly thins a point array to at most `max_points`.
pub fn stride_down(points: &[f32], max_points: usize) -> Vec<f32> {
    let count = points.len() / 3;
    if count <= max_points || max_points == 0 {
        return points.to_vec();
    }
    let step = count.div_ceil(max_points);
    let mut out = Vec::with_capacity((count / step + 1) * 3);
    let mut i = 0;
    while i < count {
        out.extend_from_slice(&points[i * 3..i * 3 + 3]);
        i += step;
    }
    out
}

/// Per-point surface normals from the smallest principal component of the
/// neighbourhood within `radius`.
///
/// Orientation is left arbitrary: point-to-plane ICP squares the plane
/// residual, so a flipped normal costs nothing, and orienting them
/// consistently across a scan is a much harder problem than the solver needs
/// solved. Points with too few neighbours get a zero normal, which `icp` reads
/// as "no usable plane here".
pub fn estimate_normals(grid: &PointGrid, radius: f64, min_neighbors: usize) -> Vec<f32> {
    let points = &grid.points;
    let count = points.len() / 3;
    let mut normals = vec![0.0f32; count * 3];
    let mut neighborhood: Vec<usize> = Vec::with_capacity(48);

    for i in 0..count {
        let (x, y, z) = (
            points[i * 3] as f64,
            points[i * 3 + 1] as f64,
            points[i * 3 + 2] as f64,
        );
        grid.neighbors(x, y, z, radius, 48, &mut neighborhood);
        if neighborhood.len() < min_neighbors {
            continue;
        }

        let n = neighborhood.len() as f64;
        let mut centre = [0.0f64; 3];
        for &index in &neighborhood {
            for axis in 0..3 {
                centre[axis] += points[index * 3 + axis] as f64;
            }
        }
        for value in centre.iter_mut() {
            *value /= n;
        }

        let (mut xx, mut xy, mut xz, mut yy, mut yz, mut zz) = (0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
        for &index in &neighborhood {
            let dx = points[index * 3] as f64 - centre[0];
            let dy = points[index * 3 + 1] as f64 - centre[1];
            let dz = points[index * 3 + 2] as f64 - centre[2];
            xx += dx * dx;
            xy += dx * dy;
            xz += dx * dz;
            yy += dy * dy;
            yz += dy * dz;
            zz += dz * dz;
        }

        let (_, vectors) = symmetric_eigen::<3>(&[xx, xy, xz, xy, yy, yz, xz, yz, zz]);
        let normal = vectors[2];
        let length = (normal[0] * normal[0] + normal[1] * normal[1] + normal[2] * normal[2]).sqrt();
        if length > 1e-12 {
            for axis in 0..3 {
                normals[i * 3 + axis] = (normal[axis] / length) as f32;
            }
        }
    }
    normals
}

/// Robust working diameter of a cloud: twice the 90th-percentile distance from
/// its median centre.
///
/// The bounding box is the wrong scale reference for a station scan. Half the
/// points of a real scan sit within a few metres of the tripod while a thin
/// tail of long-range returns stretches the box to a hundred metres or more, so
/// any parameter derived from the box ends up an order of magnitude too coarse
/// for the geometry two stations actually share.
pub fn robust_extent(points: &[f32], max_samples: usize) -> f64 {
    let count = points.len() / 3;
    if count == 0 {
        return 0.0;
    }
    let stride = count.div_ceil(max_samples.max(1)).max(1);
    let mut axes: [Vec<f64>; 3] = [Vec::new(), Vec::new(), Vec::new()];
    let mut i = 0;
    while i < count {
        for axis in 0..3 {
            axes[axis].push(points[i * 3 + axis] as f64);
        }
        i += stride;
    }

    let mut centre = [0.0f64; 3];
    for axis in 0..3 {
        let mut sorted = axes[axis].clone();
        sorted.sort_by(f64::total_cmp);
        centre[axis] = sorted[sorted.len() / 2];
    }

    let mut radii: Vec<f64> = (0..axes[0].len())
        .map(|k| {
            let dx = axes[0][k] - centre[0];
            let dy = axes[1][k] - centre[1];
            let dz = axes[2][k] - centre[2];
            (dx * dx + dy * dy + dz * dz).sqrt()
        })
        .collect();
    radii.sort_by(f64::total_cmp);
    (radii[(radii.len() as f64 * 0.9) as usize] * 2.0).max(1e-6)
}
